File: problems/2946-matrix-similarity-after-cyclic-shifts/solution.c
bool areSimilar(int** mat, int matSize, int* matColSize, int k) {
    int m = matSize;
    int n = matColSize[0];
    int** original = (int**)malloc(m * sizeof(int*));
    for (int i = 0; i < m; i++) {
        original[i] = (int*)malloc(n * sizeof(int));
        for (int j = 0; j < n; j++) {
            original[i][j] = mat[i][j];
        }
    }

    for (int step = 0; step < k; step++) {
        int** temp = (int**)malloc(m * sizeof(int*));
        for (int i = 0; i < m; i++) {
            temp[i] = (int*)malloc(n * sizeof(int));
        }
        for (int i = 0; i < m; i++) {
            if (i % 2 == 0) {
                for (int j = 0; j < n; j++) {
                    temp[i][j] = mat[i][(j + 1) % n];
                }
            } else {
                for (int j = 0; j < n; j++) {
                    temp[i][j] = mat[i][(j + n - 1) % n];
                }
            }
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                mat[i][j] = temp[i][j];
            }
            free(temp[i]);
        }
        free(temp);
    }
    bool similar = true;
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            if (mat[i][j] != original[i][j]) {
                similar = false;
                break;
            }
        }
        free(original[i]);
        if (!similar) break;
    }
    free(original);
    return similar;
}

File: problems/2946-matrix-similarity-after-cyclic-shifts/solution.java
class Solution {
    public boolean areSimilar(int[][] mat, int k) {
        int m = mat.length;
        int n = mat[0].length;
        int[][] original = new int[m][n];
        for (int i = 0; i < m; i++) {
            System.arraycopy(mat[i], 0, original[i], 0, n);
        }
        int[][] temp = new int[m][n];
        for (int i = 0; i < k; i++) {
            for (int j = 0; j < m; j++) {
                System.arraycopy(mat[j], 0, temp[j], 0, n);
            }
            for (int j = 0; j < m; j++) {
                if (j % 2 == 0) {
                    int first = temp[j][0];
                    for (int l = 0; l < n - 1; l++) {
                        mat[j][l] = temp[j][l + 1];
                    }
                    mat[j][n - 1] = first;
                } else {
                    int last = temp[j][n - 1];
                    for (int l = n - 1; l > 0; l--) {
                        mat[j][l] = temp[j][l - 1];
                    }
                    mat[j][0] = last;
                }
            }
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] != original[i][j]) {
                    return false;
                }
            }
        }
        return true;
    }
}